Added color configuration to each of the cells

mtgomes opened the pull request:
Using the same strategy as the global colouring, added PARTag that contains the title and the color definition. This will allow for, each of the labels, to have their own color scheme. 

Updated the example to reflect this and added some commented code to better display the different cases.

mtgomestest committed:
Changed the tag to get the colours from a delegate to allow for per-cell configuration of color.

mtgomestest committed:
Changed the tagPicker to contain an indexPath and not an index.

mtgomestest committed:
Removed the delegate and added the color reference per tag to allow custom colours.

mtgomestest committed:
Fixed crash and bug in colouring.

mtgomestest committed:
Fixed issues in display

mtgomestest committed:
Merge branch 'release/ReleaseToProd'

mtgomestest committed:
Removed code that was not necessary anymore.
Added the validation if the color scheme exists, otherwise uses the default one.
Improved the example.

mtgomestest committed:
Merge branch 'release/ReleaseToProd'

mtgomes committed:
Update README.md

Changed the documentation to reflect the changes.

mtgomestest committed:
Merge branch 'hotfix/ChangedDocumentation' into develop

paulrolfe commented:
Thanks, this looks great at first glance. I should have time to review it more thoroughly later today.

mtgomes commented:
No worries. I'm a Swift guy so apologies for the potential mistakes.

mtgomestest committed:
Added a tagID as a way to identify the tag other than the title.

paulrolfe commented on PARTagPicker/Classes/PARTagPickerViewController.h:
```diff
@@ -47,12 +49,12 @@ typedef NS_ENUM(NSUInteger, PARTagPickerVisibilityState) {
  *  Array of strings.
  */
-@property (nonatomic, strong) NSMutableArray *chosenTags;
+@property (nonatomic, strong) NSMutableArray<PARTag *> *chosenTags;
```
For backwards compatibility, I'd like leave this without the generics. Objective-C users would be fine, but swift users would be confined to passing in PARTags's I believe. So, I'd like consumers of this library to still be able to pass in strings, without creating a `PARTag`.

Then in the setter in the implementation we can turn each object into a PARTag if it's not already one.

paulrolfe commented on PARTagPicker/Classes/PARTagPickerViewController.h:
```diff
@@ -47,12 +49,12 @@ typedef NS_ENUM(NSUInteger, PARTagPickerVisibilityState) {
  *  Array of strings.
  */
-@property (nonatomic, strong) NSArray *allTags;
+@property (nonatomic, strong) NSArray<PARTag *> *allTags;
```
For backwards compatibility, I'd like leave this without the generics. Objective-C users would be fine, but swift users would be confined to passing in PARTags's I believe. So, I'd like consumers of this library to still be able to pass in strings, without creating a `PARTag`.

Then in the setter in the implementation we can turn each object into a PARTag if it's not already one.

paulrolfe replied:
Those setters would then look something like:

```
- (void)setAllTags:(NSArray *)allTags {
    NSArray<PARTag *> *convertedTags = [self convertToTagObjects: allTags];
    _allTags = convertedTags;
    self.availableTags = [_allTags mutableCopy];
    [self transferChosenTagsWithNewAllTags];
    [self.availableTags removeObjectsInArray: _chosenTags];
    [self filterTagsFromSearchString];
}

- (void)setChosenTags:(NSMutableArray *)chosenTags {
    NSMutableArray<PARTag *> *convertedTags = [self convertToTagObjects: chosenTags];
    _chosenTags = convertedTags;
    [self.chosenTagCollectionView reloadData];
    self.availableTags = [_allTags mutableCopy];
    [self.availableTags removeObjectsInArray: _chosenTags];
    [self filterTagsFromSearchString];
}

- (NSMutableArray *)convertToTagObjects:(NSArray *)otherArray {
    NSMutableArray *newArray = [NSMutableArray new];
    for (id object in otherArray) {
        if ([object isKindOfClass:[PARTag class]]) {
            [newArray addObject:object];
        } else if ([object isKindOfClass:[NSString class]]) {
            PARTag *tag = [[PARTag alloc] initWithText:(NSString *)object];
            [newArray addObject:tag];
        }
    }
    return newArray;
}
```

paulrolfe commented on PARTagPicker/Classes/Tag/PARTag+Comparison.h:
```diff
@@ -0,0 +1,15 @@
+
+@interface PARTag (Comparison)
+
+- (PARTag *)similarStringFromArray:(NSArray<PARTag *> *)strings;
```
Shouldn't this be called `(PARTag *)similarTagFromArray:(NSArray<PARTag *> *)tags;`

mtgomes replied:
Yep, forgot to change the name..

paulrolfe commented on PARTagPicker/ViewController.m:
```diff
@@ -29,8 +30,38 @@ - (void)viewDidLoad {
-    self.preChosenTags = @[@"in a box", @"with a fox", @"thing 1", @"thing 2"];
+    /*self.allTags = @[@"one fish", @"two fish", @"red fish", @"blue fish", @"the cat in the hat", @"Seuss"];
+    self.preChosenTags = @[@"in a box", @"with a fox", @"thing 1", @"thing 2"];*/
+    
```
Add a comment that everything below here is an example for using PARTags instead of just NSString

paulrolfe commented on PARTagPicker/ViewController.m:
```diff
@@ -69,7 +100,7 @@ - (void)addTagPickerToView {
 }
 
-- (void)useCustomColors {
+- (PARTagColorReference *)useCustomColors {
```
I'm fine with this, but rename to just `customColors`

paulrolfe commented on PARTagPicker/main.m:
```diff
@@ -1,4 +1,4 @@
-//
+ //
```
Extra space

paulrolfe commented on PARTagPicker/Classes/Tag/PARTag.h:
```diff
@@ -0,0 +1,22 @@
+
+@property (nonatomic, strong) NSString *text;
+
+@property (nonatomic, strong) PARTagColorReference *colorReference;
```
Can you tag this with `_nullable`?

paulrolfe commented on PARTagPicker/Classes/PARTagPickerViewController.m:
```diff
@@ -155,14 +155,14 @@ - (void)filterTagsFromSearchString {
 
-- (NSString *)tagSimilarToTag:(NSString *)tag {
-    NSPredicate *pred = [NSPredicate predicateWithFormat:@"self LIKE %@", tag];
+- (PARTag *)tagSimilarToTag:(NSString *)tag {
```
Should be `tagSimilarToString: (NSString *)string`

mtgomes replied:
Exactly

paulrolfe commented on PARTagPicker/Classes/PARTagPickerViewController.m:
```diff
@@ -384,13 +412,16 @@ - (BOOL)shouldReturnFromTextFieldCollectionViewCell:(PARTextFieldCollectionViewC
         
         if (self.allowsNewTags) {
-            NSString *possibleMatch = [self tagSimilarToTag:tag];
+            PARTag *possibleMatch = [self tagSimilarToTag:tag];
```
This is actually a comment about line 409: we should call variable `tag` something like `tagString` now.

paulrolfe commented on PARTagPicker/Classes/Tag/PARTag+Comparison.m:
```diff
@@ -0,0 +1,19 @@
+
+@implementation PARTag (Comparison)
+
+- (PARTag *)similarStringFromArray:(NSArray<PARTag *> *)strings{
```
`- (PARTag *)similarTagFromArray:(NSArray<PARTag *> *)tags { ... }`

paulrolfe commented on PARTagPicker/Classes/Tag/PARTag.m:
```diff
@@ -0,0 +1,35 @@
+    self = [super init];
+    if (self) {
+        self.text = text;
+        //self.colorReference = [[PARTagColorReference alloc] initWithDefaultColors];
```
delete line

paulrolfe commented on PARTagPicker/ViewController.m:
```diff
@@ -29,8 +30,38 @@ - (void)viewDidLoad {
+    NSMutableArray<PARTag *> *allTagsCreated = [[NSMutableArray alloc] initWithCapacity:allTags.count];
+    NSMutableArray<PARTag *> *allTagsPreChosen = [[NSMutableArray alloc] initWithCapacity:preChosenTags.count];
+    
+    __weak ViewController *weakSelf = self;
```
This isn't being used, I think we should remove it.

mtgomes replied:
I left it because of the code inside the block, but indeed that should be commented also.

paulrolfe commented on PARTagPicker/ViewController.m:
```diff
@@ -84,11 +115,11 @@ - (void)useCustomColors {
     myColors.highlightedTagTextColor = [UIColor blackColor];
     
-    self.tagPicker.tagColorRef = myColors;
+    //self.tagPicker.tagColorRef = myColors;
```
delete commented line

paulrolfe commented on README.md:
```diff
@@ -31,7 +31,9 @@ Include `#import <PARTagPicker/PARTagPickerViewController.h>`.
 The basic setup is to create an instance of `PARTagPickerViewController`, add it's view, and then add the controller as a child view controller.
 
-The tags are passed in as `NSString` objects in an array. Anytime you change the `allTags` array, any existing `chosenTags` are updated to use references to the new strings if they contain a match in the updated array. Otherwise they continue to point to the old strings.
+The tags are passed in as `PARTag` objects in an array. Anytime you change the `allTags` array, any existing `chosenTags` are updated to use references to the new objects if they contain a match in the updated array. Otherwise they continue to point to the old objects.
```
... passed in as `PARTag` or `NSString` objects...

paulrolfe commented on PARTagPicker/Classes/PARTagPickerViewController.m:
```diff
@@ -295,16 +296,36 @@ - (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cell
         PARTagCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:PARTagCollectionViewCellIdentifier forIndexPath:indexPath];
-        cell.tagColorRef = self.tagColorRef;
-        NSString *tag;
+        //cell.tagColorRef = self.tagColorRef;
```
remove commented line

paulrolfe commented on PARTagPicker/Classes/PARTagPickerViewController.h:
```diff
@@ -47,12 +49,12 @@ typedef NS_ENUM(NSUInteger, PARTagPickerVisibilityState) {
 /**
  *  Array of strings.
```
Array of `NSString`s or `PARTag`s

paulrolfe reviewed (changes requested):
Yeah, I've been out of Obj-C for a while myself.

Overall I'm just concerned with not breaking this library for people already interacting with it via strings. So a lot of my suggestions are around that. Unfortunately it will be impossible to keep all interactions the same when using these new tags. 

For instance, using my suggestions, when you access the tag picker's `chosenTags` array after a delegate callback, it's going to be an array of `PARTags` now. I'd love to find a solution for that as well. It might mean tracking the tags with some internal variable and leaving `chosenTags` with whatever types it was set to externally. 

I still need to think on that problem a little.

mtgomes commented:
Good suggestions overall. In the meantime I've created another version with an ID in the tag and some other minor changes. Will integrate your comments and issue another pull request. 

I understand your concerns. Let me think about it further and comment back if I find a decent solution..

mtgomestest committed:
Added a global configuration file

mtgomestest committed:
Configuration folder now in the correct place

mtgomestest committed:
Override the tag comparison to allow for better object compassion.

mtgomestest committed:
Added tag reset. If the user wants to reset both allTags and chosenTags with already defined objects it would not work properly because the setter for allTags will filter by chosenTags before the tags are actually set.
Fixed an issue where the filter would not be done properly.
Added tag comparison. For now the ID is the main comparison but it would make sense to compare everyhing .

Commit:
Merge branch 'release/ReleaseToProd'

paulrolfe commented:
@mtgomes  I saw a notification about some new commits. You want me to take another look at this PR when I get a chance?
